File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:

        left,right = 0, len(nums)-1

        while left<=right:
            
            mid = (left+right)//2
            if nums[mid] == target:
                return mid

            if nums[left] < nums[mid] and target >= nums[left] and target < nums[mid]:
                right = mid - 1
            
            elif nums[right] > nums[mid] and target > nums[mid] and target <= nums[right]:
                left = mid + 1
            
            elif nums[left] > nums[mid]:
                right = mid - 1
            
            else: 
                left = mid + 1
        
        return -1


        

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        
        int l = 0;
        int h = nums.size()-1;

        while(l<=h) {
            int mid = l + (h-l)/2;

            if(nums[mid] == target)return mid;

            if(nums[l] <= nums[mid]) {
                // upercave 
                // left part always sorted so compare here
                if(target >= nums[l] and target < nums[mid]) {
                    // discard left part
                    h = mid - 1;
                }
                else {
                    l = mid+1;
                }
            }
            else {
                // lower carve
                // right part always sorted so compare here

                if(target > nums[mid] and target <= nums[h]) {
                    l = mid+1;
                }
                else {
                    h = mid - 1;
                }
            }

            
        }
        return -1;
    }
};